core/unit_test/TestMathematicalSpecialFunctions.hpp

```diff
@@ -1901,8 +1933,653 @@ struct TestComplexBesselH2Function {
+    h_ref_cbj8(3) = Kokkos::complex<double>(4.204904284127386e+258,
+                                            -3.962917327640782e+258);
+    h_ref_cbj8(4) =
+        Kokkos::complex<double>(8.523650188274974e+84, -2.800930999646528e+84);
+
+    h_ref_cbj9(0) =
+        Kokkos::complex<double>(-19.74435372008063, 36.71222568036936);
+    h_ref_cbj9(1) = Kokkos::complex<double>(2.495433941009553e-02, 0.0);
+    h_ref_cbj9(2) =
+        Kokkos::complex<double>(-1.302704966918034e+85, 5.820333237321211e+84);
+    h_ref_cbj9(3) =
+        Kokkos::complex<double>(3.897668823461053e+258, 4.159215325012759e+258);
+    h_ref_cbj9(4) =
+        Kokkos::complex<double>(2.865837445663514e+84, 8.486616730935941e+84);
+
+    for (int i = 0; i < N; i++) {
+      EXPECT_LE(Kokkos::abs(h_cbj8(i) - h_ref_cbj8(i)),
+                Kokkos::abs(h_ref_cbj8(i)) * 1e-13);
+      EXPECT_LE(Kokkos::abs(h_cbj9(i) - h_ref_cbj9(i)),
+                Kokkos::abs(h_ref_cbj9(i)) * 1e-13);
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int& /*i*/) const {
+    d_cbj8(0) = Kokkos::Experimental::cyl_bessel_j(8, d_z(0));
+    d_cbj8(1) = Kokkos::Experimental::cyl_bessel_j(8, d_z(1));
+    d_cbj8(2) = Kokkos::Experimental::cyl_bessel_j(8, d_z(2));
+    d_cbj8(3) = Kokkos::Experimental::cyl_bessel_j(8, d_z(3));
+    d_cbj8(4) = Kokkos::Experimental::cyl_bessel_j(8, d_z(4));
+
+    d_cbj9(0) = Kokkos::Experimental::cyl_bessel_j(9, d_z(0));
+    d_cbj9(1) = Kokkos::Experimental::cyl_bessel_j(9, d_z(1));
+    d_cbj9(2) = Kokkos::Experimental::cyl_bessel_j(9, d_z(2));
+    d_cbj9(3) = Kokkos::Experimental::cyl_bessel_j(9, d_z(3));
+    d_cbj9(4) = Kokkos::Experimental::cyl_bessel_j(9, d_z(4));
+  }
+};
+
+template <class ExecSpace>
+struct TestBesselCalsWithDeduction {
+  using ViewType = Kokkos::View<Kokkos::complex<double>*, ExecSpace>;
+  using HostViewType =
+      Kokkos::View<Kokkos::complex<double>*, Kokkos::HostSpace>;
+
+  void testit() {
+    Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpace>(0, 1), *this);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int& /*i*/) const {
+    Kokkos::Experimental::cyl_bessel_j(6, Kokkos::complex<double>(1.0, 1.0));
+    Kokkos::Experimental::cyl_bessel_i(6, Kokkos::complex<double>(1.0, 1.0));
+    Kokkos::Experimental::cyl_bessel_k(6, Kokkos::complex<double>(1.0, 1.0));
+    Kokkos::Experimental::cyl_bessel_y(6, Kokkos::complex<double>(1.0, 1.0));
+    Kokkos::Experimental::cyl_bessel_h1(6, Kokkos::complex<double>(1.0, 1.0));
+    Kokkos::Experimental::cyl_bessel_h2(6, Kokkos::complex<double>(1.0, 1.0));
+
+    // Call to vendor functions of possible and ret real type
+    Kokkos::Experimental::cyl_bessel_j(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_i(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_k(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_y(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_h1(6, 1.0);
+    Kokkos::Experimental::cyl_bessel_h2(6, 1.0);
```

Comment:
what is being tested here? just that it compiles?

Reply:
Yes, only that it properly deducing templates, i can also add tests for the numerical value of those

Reply:
Please do.

Reply:
I added tests for higher orders for both real and complex arguments, and I was lucky to find out divergene between results taken from octave and julia. So I need to rethink implementation, if anyone have any ideas i will be happy to hear them